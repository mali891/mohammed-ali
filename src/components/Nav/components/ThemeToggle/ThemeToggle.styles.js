import { css } from '@emotion/core'
import { COLOURS } from '../../../../styles/settings'

export const styles = css`
  display: flex;
  height: 3.75rem;

  .c-theme-toggle__input {
    height: 0;
    width: 0;
    visibility: hidden;
  }

  .c-theme-toggle__label {
    cursor: pointer;
    text-indent: -9999px;
    width: 7.5rem;
    height: 3.75rem;
    background: grey;
    display: block;
    border-radius: 10rem;
    position: relative;

    &:active {
      .c-theme-toggle__toggle {
        width: 4.5rem;
      }
    }
  }

  .c-theme-toggle__toggle {
    content: '';
    position: absolute;
    top: 0.1875rem;
    left: 0.1875rem;
    width: 3.375rem;
    height: 3.375rem;
    background: #fff;
    border-radius: 9rem;
    transition: 0.3s;
  }

  .c-theme-toggle__input:checked + .c-theme-toggle__label {
    background: ${COLOURS.GRADIENT_THREE.FIRST};
  }

  .c-theme-toggle__input:checked + .c-theme-toggle__toggle {
    left: calc(100% - 0.1875rem);
    transform: translateX(-100%);
  }
`
