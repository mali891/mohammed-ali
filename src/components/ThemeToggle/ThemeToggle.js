import React from 'react'

const ThemeToggle = () => (
  <label htmlFor="theme-toggle">
    <input id="theme-toggle" name="theme" type="checkbox" />
  </label>
)

export default ThemeToggle
