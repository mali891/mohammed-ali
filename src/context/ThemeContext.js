import React, { createContext, useState } from 'react'
import PropTypes from 'prop-types'

const { Provider, Consumer } = createContext()

const ThemeProvider = ({ children }) => {
  const [theme, setTheme] = useState('dark')

  return <Provider value={{ theme, setTheme }}>{children}</Provider>
}

ThemeProvider.propTypes = {
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node])
}

export { ThemeProvider, Consumer as ThemeConsumer }
