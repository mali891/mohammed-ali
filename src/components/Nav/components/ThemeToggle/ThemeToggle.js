import React from 'react'

import { ThemeConsumer } from '../../../../context/ThemeContext'
import { styles } from './ThemeToggle.styles'
import { Sun, Moon } from '../../../../icons'

const toggleTheme = (theme, setTheme) => (theme === 'dark' ? setTheme('light') : setTheme('dark'))
const handleKeyDown = (e, theme, setTheme) => e.keyCode === 13 && toggleTheme(theme, setTheme)

const ThemeToggle = () => (
  <ThemeConsumer>
    {({ theme, setTheme }) => (
      <div
        css={styles}
        className={`c-theme-toggle c-theme-toggle--${theme}`}
        aria-label="Toggle theme and colour contrast"
      >
        <input
          className="c-theme-toggle__input"
          type="checkbox"
          id="theme-toggle"
          checked={theme === 'light'}
          readOnly
        />
        <label
          className="c-theme-toggle__label"
          onClick={() => toggleTheme(theme, setTheme)}
          onKeyDown={e => handleKeyDown(e, theme, setTheme)}
          htmlFor="theme-toggle"
          tabIndex={0}
        >
          <div className="c-theme-toggle__toggle">
            {theme === 'light' ? <Sun className="c-theme-toggle__icon" /> : <Moon className="c-theme-toggle__icon" />}
          </div>
        </label>
      </div>
    )}
  </ThemeConsumer>
)

export default ThemeToggle
