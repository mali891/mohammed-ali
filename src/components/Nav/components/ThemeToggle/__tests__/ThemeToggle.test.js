import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'

import ThemeToggle from '../ThemeToggle'
import { ThemeProvider } from '../../../../../context/ThemeContext'

describe('ThemeToggle', () => {
  it('renders correctly', () => {
    const { asFragment } = render(
      <ThemeProvider>
        <ThemeToggle />
      </ThemeProvider>
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('clicking the component toggles the state of the icon from moon to sun and back again', () => {
    const { asFragment, container } = render(
      <ThemeProvider theme="dark">
        <ThemeToggle />
      </ThemeProvider>
    )
    const label = container.querySelector('.c-theme-toggle__label')

    expect(container.querySelector('.icon--moon')).toBeInTheDocument()
    expect(container.querySelector('.icon--sun')).not.toBeInTheDocument()

    fireEvent.click(label)

    expect(container.querySelector('.icon--sun')).toBeInTheDocument()
    expect(container.querySelector('.icon--moon')).not.toBeInTheDocument()

    fireEvent.click(label)

    expect(container.querySelector('.icon--moon')).toBeInTheDocument()
    expect(container.querySelector('.icon--sun')).not.toBeInTheDocument()

    expect(asFragment()).toMatchSnapshot()
  })

  it('pressing enter on the component toggles the state of the icon from moon to sun and back again', () => {
    const e = { keyCode: 13 }
    const { asFragment, container } = render(
      <ThemeProvider theme="dark">
        <ThemeToggle />
      </ThemeProvider>
    )
    const label = container.querySelector('.c-theme-toggle__label')

    expect(container.querySelector('.icon--moon')).toBeInTheDocument()
    expect(container.querySelector('.icon--sun')).not.toBeInTheDocument()

    fireEvent.keyDown(label, e)

    expect(container.querySelector('.icon--sun')).toBeInTheDocument()
    expect(container.querySelector('.icon--moon')).not.toBeInTheDocument()

    fireEvent.keyDown(label, e)

    expect(container.querySelector('.icon--moon')).toBeInTheDocument()
    expect(container.querySelector('.icon--sun')).not.toBeInTheDocument()

    expect(asFragment()).toMatchSnapshot()
  })
})
