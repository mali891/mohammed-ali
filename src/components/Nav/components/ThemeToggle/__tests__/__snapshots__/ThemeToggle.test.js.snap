// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeToggle clicking the component toggles the state of the icon from moon to sun and back again 1`] = `
<DocumentFragment>
  <div
    aria-label="Toggle theme and colour contrast"
    class="c-theme-toggle c-theme-toggle--dark"
    css="[object Object]"
  >
    <input
      class="c-theme-toggle__input"
      id="theme-toggle"
      readonly=""
      type="checkbox"
    />
    <label
      class="c-theme-toggle__label"
      for="theme-toggle"
      tabindex="0"
    >
      <div
        class="c-theme-toggle__toggle"
      >
        <svg
          class="u-anim--fade-in icon--moon c-theme-toggle__icon"
          viewBox="0 0 487.42 513.3"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M282.89.8a240.56 240.56 0 0158.41 157.42c0 133.44-108.17 241.61-241.61 241.61a240.59 240.59 0 01-98.53-21A258.81 258.81 0 00228 512.8c143 0 259-115.93 259-258.94C486.92 129.7 399.5 26 282.89.8zM228 494a240.28 240.28 0 01-179.16-80.29 260.85 260.85 0 0050.85 5c143.61 0 260.45-116.84 260.45-260.45A260.36 260.36 0 00329.9 36.52c82.45 38.93 138.17 123.38 138.17 217.34C468.07 386.25 360.36 494 228 494z"
            data-name="moon"
            fill="#FFE29F"
            stroke="#FFE29F"
            stroke-miterlimit="10"
            stroke-width="3px"
          />
        </svg>
      </div>
    </label>
  </div>
</DocumentFragment>
`;

exports[`ThemeToggle pressing enter on the component toggles the state of the icon from moon to sun and back again 1`] = `
<DocumentFragment>
  <div
    aria-label="Toggle theme and colour contrast"
    class="c-theme-toggle c-theme-toggle--dark"
    css="[object Object]"
  >
    <input
      class="c-theme-toggle__input"
      id="theme-toggle"
      readonly=""
      type="checkbox"
    />
    <label
      class="c-theme-toggle__label"
      for="theme-toggle"
      tabindex="0"
    >
      <div
        class="c-theme-toggle__toggle"
      >
        <svg
          class="u-anim--fade-in icon--moon c-theme-toggle__icon"
          viewBox="0 0 487.42 513.3"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M282.89.8a240.56 240.56 0 0158.41 157.42c0 133.44-108.17 241.61-241.61 241.61a240.59 240.59 0 01-98.53-21A258.81 258.81 0 00228 512.8c143 0 259-115.93 259-258.94C486.92 129.7 399.5 26 282.89.8zM228 494a240.28 240.28 0 01-179.16-80.29 260.85 260.85 0 0050.85 5c143.61 0 260.45-116.84 260.45-260.45A260.36 260.36 0 00329.9 36.52c82.45 38.93 138.17 123.38 138.17 217.34C468.07 386.25 360.36 494 228 494z"
            data-name="moon"
            fill="#FFE29F"
            stroke="#FFE29F"
            stroke-miterlimit="10"
            stroke-width="3px"
          />
        </svg>
      </div>
    </label>
  </div>
</DocumentFragment>
`;

exports[`ThemeToggle renders correctly 1`] = `
<DocumentFragment>
  <div
    aria-label="Toggle theme and colour contrast"
    class="c-theme-toggle c-theme-toggle--dark"
    css="[object Object]"
  >
    <input
      class="c-theme-toggle__input"
      id="theme-toggle"
      readonly=""
      type="checkbox"
    />
    <label
      class="c-theme-toggle__label"
      for="theme-toggle"
      tabindex="0"
    >
      <div
        class="c-theme-toggle__toggle"
      >
        <svg
          class="u-anim--fade-in icon--moon c-theme-toggle__icon"
          viewBox="0 0 487.42 513.3"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M282.89.8a240.56 240.56 0 0158.41 157.42c0 133.44-108.17 241.61-241.61 241.61a240.59 240.59 0 01-98.53-21A258.81 258.81 0 00228 512.8c143 0 259-115.93 259-258.94C486.92 129.7 399.5 26 282.89.8zM228 494a240.28 240.28 0 01-179.16-80.29 260.85 260.85 0 0050.85 5c143.61 0 260.45-116.84 260.45-260.45A260.36 260.36 0 00329.9 36.52c82.45 38.93 138.17 123.38 138.17 217.34C468.07 386.25 360.36 494 228 494z"
            data-name="moon"
            fill="#FFE29F"
            stroke="#FFE29F"
            stroke-miterlimit="10"
            stroke-width="3px"
          />
        </svg>
      </div>
    </label>
  </div>
</DocumentFragment>
`;
